### PR TITLE
fix(select): resolve issue with Select component closing parent Dialog on escape key press - FE-5624

### LINKED ISSUE #5716: Press escape to dismiss dropdown on sidebar
### Current behaviour

This is a minor accessibility issue
While navigating in a sidebar, once pressing the escape key to close a drop down list the whole modal is closed instead.

### Expected behaviour

It should close the current dropdown open and not the complete sidebar

### CodeSandbox or Storybook URL

https://codesandbox.io/s/sidebar-with-select-nllhe4

### JIRA Ticket (Sage Only)

_No response_

### Suggested Solution

_No response_

### Carbon Version

111.8.0

### Design Tokens Version

2.45.0

### What browsers are you seeing the problem on?

Chrome

### What Operating System are you seeing the problem on?

MacOS

### Anything else we should know?

_No response_

### Confidentiality

- [X] I confirm there is no confidential or commercially sensitive information included.

### PR DESCRIPTION
**Pull request by mkrds**
### Proposed behaviour

Fixes #5716

### Checklist

- [x] Commits follow our style guide
- [x] Related issues linked in commit messages if required
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Related docs have been updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions

The following CodeSandbox is an example of the broken behaviour.
You can see the new behaviour by looking at the version in the comment by `codesandbox[bot]`.

### CODE REVIEW

File: src/components/select/multi-select/multi-select.component.js [hunk 1]
```diff
@@ -458,10 +458,10 @@ const MultiSelect = React.forwardRef(
       [textboxRef, actualValue, updateValue]
     );
 
-    function onSelectListClose() {
+    const onSelectListClose = useCallback(() => {
       setOpenState(false);
       setFilterText("");
-    }
+    }, []);
```
Comment (grabkowski) [inline src/components/select/multi-select/multi-select.component.js:464]: could you explain why do we need to have useCallback here?

Comment (mkrds) [reply src/components/select/multi-select/multi-select.component.js:464]: `onSelectListClose` method in `MultiSelect` is passed down to `SelectList` where is referenced inside of a `handleEscapeKey` method. The `handleEscapeKey` is then passed down to `useModalManager` which references it inside of `handleClose` method which is then added to `keyup` event listener. Not wrapping it in `useCallback` would result in events being added and removed on each rerender.

**Review by grabkowski — APPROVED** [review]

**Review by nineteen88 — APPROVED** [review]

File: src/components/select/filterable-select/filterable-select.test.js [hunk 2]
```diff
@@ -1523,4 +1538,24 @@ context("Tests for Filterable Select component", () => {
       selectOptionByText("Option 1002.").should("be.visible");
     });
   });
+
+  describe("when nested inside of a Dialog component", () => {
+    it.only("should not close the Dialog when Select is closed by pressing an escape key", () => {
```
Comment (DlgSHi) [inline src/components/select/filterable-select/filterable-select.test.js:1543]:
```suggestion
    it("should not close the Dialog when Select is closed by pressing an escape key", () => {
```

File: src/components/select/multi-select/multi-select.test.js [hunk 3]
```diff
@@ -1371,4 +1389,24 @@ context("Tests for Multi Select component", () => {
       selectOptionByText("Option 1002.").should("be.visible");
     });
   });
+
+  describe("when nested inside of a Dialog component", () => {
+    it.only("should not close the Dialog when Select is closed by pressing an escape key", () => {
```
Comment (DlgSHi) [inline src/components/select/multi-select/multi-select.test.js:1394]:
```suggestion
    it("should not close the Dialog when Select is closed by pressing an escape key", () => {
```

File: src/components/select/simple-select/simple-select.test.js [hunk 4]
```diff
@@ -1221,4 +1238,24 @@ context("Tests for Simple Select component", () => {
       selectOptionByText("Option 40.").should("not.exist");
     });
   });
+
+  describe("when nested inside of a Dialog component", () => {
+    it.only("should not close the Dialog when Select is closed by pressing an escape key", () => {
```
Comment (DlgSHi) [inline src/components/select/simple-select/simple-select.test.js:1243]:
```suggestion
    it("should not close the Dialog when Select is closed by pressing an escape key", () => {
```

**Review by DlgSHi — COMMENTED** [review]
we need to remove `.only` from the tests

**Review by grabkowski — APPROVED** [review]

**Review by nineteen88 — APPROVED** [review]

**Review by ZhuoyuJin — APPROVED** [review]

**Pull request merged**